pulsar-client/src/main/java/com/yahoo/pulsar/client/impl/ConsumerImpl.java

```diff
@@ -928,6 +943,38 @@ public void redeliverUnacknowledgedMessages(List<MessageIdImpl> messageIds) {
         }
     }
 
+    private MessageIdImpl getMessageIdImpl(Message msg) {
+        MessageIdImpl messageId = (MessageIdImpl) msg.getMessageId();
+        if (messageId instanceof BatchMessageIdImpl) {
+            // messageIds contain MessageIdImpl, not BatchMessageIdImpl
+            messageId = new MessageIdImpl(messageId.getLedgerId(), messageId.getEntryId(),
+                    getPartitionIndex());
+        }
+        return messageId;
+    }
+
+    private int removeExpiredMessagesFromQueue(Set<MessageIdImpl> messageIds) {
+        int messagesFromQueue = 0;
+        Message peek = incomingMessages.peek();
+        if (peek != null) {
+            MessageIdImpl messageId = getMessageIdImpl(peek);
+            if (!messageIds.contains(messageId)) {
+                // first message is not expired, then no message is expired in queue.
+                return 0;
+            }
+
+            int queueSize = incomingMessages.size();
+            Message message = incomingMessages.poll();
+            // try not to remove elements that are added while we remove
+            while(message != null && messagesFromQueue < queueSize) {
+                messagesFromQueue++;
+                messageIds.add(getMessageIdImpl(peek));
```

Comment:
we're not adding `peek` to the set, as it does not remove it from the queue, it's just to see if the queue contains expired messages